apps/common/lib/lexical/ast.ex

```diff
@@ -42,19 +95,98 @@ defmodule Lexical.Ast do
   end
 
   def from(s) when is_binary(s) do
-    parse(s)
+    do_string_to_quoted(s)
   end
 
-  defp parse(s) when is_binary(s) do
-    case Sourceror.string_to_quoted(s, token_metadata: true, columns: true) do
-      {:ok, code, _comments} ->
-        {:ok, code}
-
-      error ->
-        error
+  @doc """
+  Returns an AST fragment from the start of the document to the given position.
+  """
+  @spec fragment(Document.t(), Position.t()) :: {:ok, t} | {:error, parse_error()}
+  def fragment(%Document{} = document, %Position{} = position) do
+    # https://github.com/elixir-lang/elixir/issues/12673#issuecomment-1592845875
+    # Note: because of the above issue: Using `cursor_context` + `container_cursor_to_quoted`
+    # can't deal with some cases like: `alias Foo.Bar, as: AnotherBar`,
+    # so we need to add a new line to make sure we can get the parrent node of the cursor
+    %{line: line} = position
+    added_new_line_position = Position.new(line + 1, 1)
+    fragment = Document.fragment(document, added_new_line_position)
+
+    case do_container_cursor_to_quoted(fragment) do
+      {:ok, quoted} ->
+        {:ok, quoted}
+
+      _error ->
+        # https://github.com/elixir-lang/elixir/issues/12673#issuecomment-1626932280
+        # NOTE: Adding new line doesn't always work,
+        # so we need to try again without adding new line
+        document_fragment = Document.fragment(document, position)
+        do_container_cursor_to_quoted(document_fragment)
     end
   end
 
+  @doc """
+  Parses the given fragment into an AST.
+  """
+  @spec fragment(String.t()) :: {:ok, t} | {:error, parse_error()}
+  def fragment(s) when is_binary(s) do
+    do_container_cursor_to_quoted(s)
+  end
+
+  @doc """
+  Returns the cursor context of the document at a position.
+  """
+  @spec cursor_context(Document.t(), Position.t()) ::
+          {:ok, cursor_context()} | {:error, :cursor_context}
+  def cursor_context(%Document{} = document, %Position{} = position) do
+    document
+    |> Document.fragment(position)
+    |> cursor_context()
+  end
+
+  @doc """
+  Returns the cursor context of the fragment.
+  """
+  @spec cursor_context(String.t()) :: {:ok, cursor_context()} | {:error, :cursor_context}
+  def cursor_context(fragment) when is_binary(fragment) do
```

Comment:
the variable name is a fragment, but its type is string, which feels strange because the fragment function above returns t

Reply:
Good points. I was copying the naming from `Code.Fragment.cursor_context`, but I agree that it's confusing given the presence of another `fragment` function. I'm inclined to just rename to `s` or `string`.